src/connections.rs

```diff
@@ -225,8 +230,17 @@ impl IncomingMessages {
                 error!("Failed to read incoming message on bi-stream: {}", err);
                 None
             }
-            Some(Ok((send, mut recv))) => match read_bytes(&mut recv).await {
-                Ok(bytes) => Some((bytes, send, recv)),
+            Some(Ok((mut send, mut recv))) => match read_bytes(&mut recv).await {
+                Ok(WireMsg::UserMsg(bytes)) => Some((bytes, send, recv)),
+                Ok(WireMsg::EndpointEchoReq) => {
+                    let message = WireMsg::EndpointEchoResp(peer_addr);
+                    message.write_to_stream(&mut send).await.ok()?;
+                    None
+                }
+                Ok(msg) => {
+                    error!("Unexpected message type: {:?}", msg);
+                    None
```

Comment:
Same as above comment, we should log the error but we shouldn't return `None`, instead read next messages by continuing the loop.